hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/signatures/SignatureVerifier.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.signatures;
+
+import com.hedera.node.app.spi.Service;
+import com.hedera.node.app.spi.key.HederaKey;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Allows either a workflow or a {@link Service} to test if a {@link HederaKey} has signed at the
+ * current point in the workflow.
+ *
+ * <p>There are three kinds of "checkpoints" where we test signatures:
+ *
+ * <ol>
+ *   <li>Before submitting or executing a top-level HAPI transaction.
+ *   <li>Before triggering the execution of a scheduled transaction.
+ *   <li>Before executing a EVM system contract.
+ * </ol>
+ *
+ * Note that workflows are solely responsible for enforcing the first checkpoint, while {@link
+ * Service} business logic must detect and enforce the second and third checkpoints.
+ */
+public interface SignatureVerifier {
+    /**
+     * Returns whether the given key has signed at the current checkpoint.
+     *
+     * @param key the key to test
+     * @return whether the key has signed
+     */
+    boolean hasSigned(@NonNull HederaKey key);
```

Comment:
I am afraid, I do not really understand how this works. Is there some documentation of the overall design?

Reply:
I will add a flow diagram when I am implementing it this week @netopyr